main.tf

```diff
@@ -1,152 +1,165 @@
 
-terraform {
-  required_version = ">= 0.12.24"
+resource "null_resource" "variable_validation" {
+  lifecycle {
+    postcondition {
+      condition       = (contains(local.provision_modes_values_list,var.provision_mode))
+      error_message   = "Invalid provision mode: ${var.provision_mode}. Valid provision modes are: ${join(", ",local.provision_modes_values_list)}"
+    }
+    postcondition {
+      condition       = (!var.deploy_networking || var.bastion_enabled)
+      error_message   = "bastion_enabled must be true if deploy_networking is true"
+    }
+    postcondition {
+      condition       = (!var.bastion_enabled || (var.bastion_enabled && var.bastion_private_key_path != ""))
+      error_message   = "bastion_private_key_path must be set if bastion_enabled is true"
+    }
+  }
 }
 
-module "base" {
-  source  = "oracle-terraform-modules/base/oci"
-  version = "1.3.1"
-
-  # general oci parameters
-  oci_base_general = local.oci_base_general
-
-  # identity
-  oci_base_provider = local.oci_base_provider
-
-  # vcn parameters
-  oci_base_vcn = local.oci_base_vcn
-
-  # bastion parameters
-  oci_base_bastion = local.oci_base_bastion
-
-  # operator server parameters
-  oci_base_operator = local.oci_base_operator
-
+data "external" "ocne_config" {
+  program = ["sh", "${path.module}/modules/ocne-provision/files/config-read.sh"]
+  query = {
+    path = var.config_file_path
+    ov = var.ocne_version
+    en = var.environment_name
+    kn = var.kubernetes_name
+  }
 }
 
-module "network" {
-  source = "./modules/network"
-
-  # general oci parameters
-  compartment_id = var.compartment_id
-  label_prefix   = var.label_prefix
-  ad_names       = module.base.ad_names
-
-  # network parameters
-  ig_route_id  = module.base.ig_route_id
-  nat_route_id = module.base.nat_route_id
-  netnum       = var.netnum
-  newbits      = var.newbits
-  vcn_id       = module.base.vcn_id
+resource "null_resource" "ocne_config_validation" {
+  depends_on = [data.external.ocne_config]
+  lifecycle {
+    postcondition {
+      condition       = (length(data.external.ocne_config.result.error) == 0)
+      error_message   = "${data.external.ocne_config.result.error}"
+    }
+  }
 }
 
-module "master" {
-  source = "./modules/master"
-
-  # general oci parameters
-  compartment_id = var.compartment_id
-  label_prefix   = var.label_prefix
-  ad_names       = module.base.ad_names
-
-  # networking parameters
-  nsg_id       = module.network.nsg_ids["master"]
-  subnet_id    = module.network.subnet_ids["masters"]
-  subnet_label = module.network.master_subnet_dns_label
-
-  # master compute parameters
-  master_image_id     = var.master_image_id
-  master_shape        = var.master_shape
-  master_size         = var.master_size
-  master_upgrade      = var.master_package_upgrade
-  ssh_public_key      = ""
-  ssh_public_key_path = var.ssh_public_key_path
-  timezone            = var.master_timezone
-
+module "oci-ocne-vcn" {
+  source     = "./modules/oci-ocne-vcn"
```

Comment:
please use the vcn module https://registry.terraform.io/modules/oracle-terraform-modules/vcn/oci/latest

Reply:
This module is renamed to oci-ocne-network, which sets up VCN, subnets, as well as, bastion.
I did remove terraform-oci-vcn module and use oracle-terraform-modules/vcn/oci in https://github.com/oracle-terraform-modules/terraform-oci-olcne/blob/58cade472a834306e446e9820642a6482f694281/modules/oci-ocne-network/main.tf#L6

Reply:
cool thank you.

Reply:
I see you are also embedding the bastion module in network. We tried this approach in the oke module and we were using the base module. It caused us endless grief especially when we had to upgrade, add a feature or a fix in the submodule. We had to cut a release every time. 

Can I suggest you keep submodules at 1 level as much a possible?  It's not that it's bad, it'll be easier for you to maintain going forward

Reply:
Moved the bastion module from network to https://github.com/guoyongzhang/terraform-oci-olcne/blob/gz/uptake_internal_ocne_tf/main.tf#L61